### DISCUSSION TITLE
How to query deep nested data?

### DISCUSSION THREAD

**Discussion by siddhesh-nalawade** [cofu-app/cbl-dart#409] [Q&A] [answered]
I have nested json document which I want to store in couchbase lite. Also I want to query the data from it but I found that some of the important features are not supported like UNNEST or subqueries. How should I perform queries on my data?

The structure of data is as follows:

```json
{
  "mainCategoryName": "",
  "options": [
    {
      "name": "",
      "details": [
        {
          "title": "",
          "detailsArray": [
            {
              "BackgroundColour": "",
              "name": "",
              "information": []
            }
          ]
        }
      ]
    }
  ],
  "definition": [],
  "id": "",
  "symptomName": ""
}
```

**Comment by blaugold** [comment, marked as answer]
Hey @Siddhesh-Nalawade, great question.

First, you might find the [reference for SQL++](https://docs.couchbase.com/couchbase-lite/current/c/query-n1ql-mobile.html#lbl-exp-any) helpful.

Specifically for querying data in arrays, the [`ANY` and `EVERY` expressions](https://docs.couchbase.com/couchbase-lite/current/c/query-n1ql-mobile.html#lbl-exp-any) are useful.

The query below returns every document with at least one matching `option`. Only `options` with the `name` `"foo"` and at least one `detail` with the `title` `"bar"` will match.

```sql
SELECT *
FROM _
WHERE
  ANY option
    IN options
    SATISFIES
      option.name = "foo" AND
      ANY detail
      IN option.details
      SATISFIES
        detail.title = "bar"
      END
  END
```

You can create a SQL++ query through [`Query.fromN1QL()`](https://pub.dev/documentation/cbl/latest/cbl/Query/fromN1ql.html). N1QL is an earlier name for SQL++.

**Reply by siddhesh-nalawade** [reply]
@blaugold Thanks for your reply. Actually I am creating mobile app using flutter which will store data offline. Also I want to implement full text search feature. So I am thinking to use Couchbase Lite. How can I create index for ''information" field and search in it?

**Reply by blaugold** [reply]
Here is the [reference](https://docs.couchbase.com/couchbase-lite/current/c/query-n1ql-mobile.html#lbl-func-fts) for full-text search functions, and here are the docs for the [pattern matching syntax](https://docs.couchbase.com/couchbase-lite/current/c/fts.html#pattern-matching-formats).

```dart
// This creates a full-text index named "informationFti" that indexes the 
// `information` field of all documents.
await db.createIndex("informationFts", FullTextIndexConfiguration(['information']));

// This query returns all documents whose `information` field contains the word `foo`.
final query = await Query.fromN1ql(db, '''
SELECT *
FROM _
WHERE MATCH (informationFti, "foo")
''');
```
